### PR TITLE
Update/unify jest@27 across all packages

### PR DESCRIPTION
**Pull request by tomalec**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Unifies all jest versions used across the repo to `^27`. First to update the legacy ones, and keep them in sync to avoid conflicts.

This is an extension of/attempt to make https://github.com/woocommerce/woocommerce/pull/33827 pass all the checks.

### How to test the changes in this Pull Request:

1. Run automated tests for all packages

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Review by ObliviousHarmony — CHANGES_REQUESTED** [review]
This seems to be failing because of `ts-jest` not being updated too. Make sure to also update anything that depends on a specific version of Jest to a compatible version!

**Comment by tomalec** [comment]
> This seems to be failing because of `ts-jest` not being updated too. Make sure to also update anything that depends on a specific version of Jest to a compatible version!

Could you tell me where it is not updated?
All the results for `"ts-jest"` search, gives: `"ts-jest": "^27.1.3"` or `"ts-jest": "^27"` which should be compatible with Jest 27

**Comment by tomalec** [comment]
I investigated it a bit deeper, and it seems that `import { mocked } from 'ts-jest/utils';` is problematic here. After I removed this import and naively all its uses, `tsc` build no longer fails.
https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/api/src/repositories/rest/__tests__/shared.spec.ts#L1

According to [docs](https://kulshekhar.github.io/ts-jest/docs/27.1/guides/test-helpers/):
> This function is now deprecated and will be removed in 28.0.0. The function has been integrated into jest-mock package as a part of Jest 27.4.0, see https://github.com/facebook/jest/pull/12089. Please use the one from jest-mock instead.

So maybe something got less stable by `27`. I'll try to replace mocked and see how it goes.

------

However, @ObliviousHarmony I would still appreciate it if you could point out where I missed updating `ts-jest`, or why you think that's the case.

**Comment by tomalec** [comment]
The build runs successfully on my local, but I don't know why the GH actions are not fired any longer here.

Edit: for future reference, the branch needs no conflicts for actions to run.

**Comment by ObliviousHarmony** [comment]
Sorry @tomalec, I saw that you didn't change any `ts-jest` versions and wrongfully assumed that they had already been matched to the `jest` versions. Thank you for clearing that confusion up and resolving the problems. Since we're going through the troubling of synchronizing all `jest` versions, we should add them to the `syncpack` version constraints. In #34397 we integrated a tool to make sure that certain dependencies are always synchronized and this seems like a great candidate.

I'd also like to cc @woocommerce/solaris here because this includes updates to the `jest` version in their E2E packages. We need to have their approval as well before we can move forward. Since this updates the `e2e-environment` package, there may be some downstream implications with the test runner, but, I'm not sure.

**Comment by tomalec** [comment]
>  I saw that you didn't change any `ts-jest` versions

I did at https://github.com/woocommerce/woocommerce/pull/34322/files#diff-a135adc74d2c7f8a5a69fc48fea9a7c400a015bacb97139b752622d114f3ecfbR55

But, I hope it's clear now.

>Since we're going through the troubling of synchronizing all jest versions, we should add them to the syncpack version constraints. In https://github.com/woocommerce/woocommerce/pull/34397 we integrated a tool to make sure that certain dependencies are always synchronized and this seems like a great candidate.


Wouldn't it be better to make it a separate PR? AFAIK to do so I'd need to change existing settings.
Like change `filter` property https://github.com/woocommerce/woocommerce/pull/34397/files#diff-aac9fa94496fd53bd351b6dbf9fa82cfd55e6abcff3b0f27169590744b2c5747R3
Which made the entire syncpack setup very react-only, and makes impossible to just add another `versionGroup` for `jest`, to make it work.

That could affect React dependencies, and that enlarges the scope of this PR heavily. Not to mention, that my initial intent was to just bump it make consistent around E2E.

In general, I believe having smaller PRs makes them easier to review

**Comment by tomalec** [comment]
I made a separate PR to add jest to syncpack, and unify the version more precisely on `^27.5.1` https://github.com/woocommerce/woocommerce/pull/34468 - it can be merged before or after this one.

I'd prefer after, to speed up merging this, as I suspect this may solve/help to debug broken PR checks for things like https://github.com/woocommerce/woocommerce/pull/33901

**Comment by tammullen** [comment]
> I'd also like to cc https://github.com/orgs/woocommerce/teams/solaris here because this includes updates to the jest version in their E2E packages. We need to have their approval as well before we can move forward. Since this updates the e2e-environment package, there may be some downstream implications with the test runner, but, I'm not sure.

cc @woocommerce/masamune for `e2e-environment`

File: packages/js/api-core-tests/changelog/update-jest
Comment (tammullen) [inline packages/js/api-core-tests/changelog/update-jest]: I don't think this changelog entry is needed as the update was already made in https://github.com/woocommerce/woocommerce/pull/34345/ and the changelog updated as part of that
https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/api-core-tests/CHANGELOG.md

Comment (tomalec) [reply packages/js/api-core-tests/changelog/update-jest]: Isn't that what JetPack Changelogger flow is for, to avoid such conflicts? Also, this PR and file were made before https://github.com/woocommerce/woocommerce/pull/34345.

Should I remove this file? Would PR check be satisfied, given I change something in `api-core-test` w/o providing a changelog note?

Comment (tammullen) [reply packages/js/api-core-tests/changelog/update-jest]: > Should I remove this file? Would PR check be satisfied, given I change something in api-core-test w/o providing a changelog note?

I think so as looking further into it I don't believe `api-core-tests` uses the changelogger, it was excluded as it already had an existing workflow for updating the changelog.

Comment (tomalec) [reply packages/js/api-core-tests/changelog/update-jest]: Thanks, for pointing that out and the explanations :) 
Removed in https://github.com/woocommerce/woocommerce/pull/34322/commits/2033bb24a7017b2b56f6752be828b09894271ce8

**Comment by MrJnrman** [comment]
Thanks for the work here @tomalec! I'm happy to see this PR passing in our CI. 

I pulled this branch locally and tried out the `api-core-tests` and `e2e-core-tests` on the `e2e-environment` package. The API tests worked fine, but the E2E the tests are hanging when ran in headless mode and when ran in dev move, I see the error below:

```
Error: Jest: Got error running globalSetup - /Users/Code/woocommerce/node_modules/.pnpm/jest-environment-puppeteer@5.0.4/node_modules/jest-environment-puppeteer/setup.js, reason: ErrorEvent {
  "error": [Error: socket hang up],
  "message": "socket hang up",
  "target": WebSocket {
    "_binaryType": "nodebuffer",
    "_closeCode": 1006,
    "_closeFrameReceived": false,
    "_closeFrameSent": false,
    "_closeMessage": "",
    "_closeTimer": null,
    "_events": Object {
      "error": [Function onError],
      "open": [Function onOpen],
    },
    "_eventsCount": 2,
    "_extensions": Object {},
    "_isServer": false,
    "_maxListeners": undefined,
    "_receiver": null,
    "_redirects": 0,
    "_req": null,
    "_sender": null,
    "_socket": null,
    "protocol": "",
    "readyState": 3,
    "url": "ws://127.0.0.1:54223/devtools/browser/28a4692c-9e70-4fc8-80b4-6233022acd64",
    Symbol(kCapture): false,
  },
  "type": "error",
}
    at runGlobalHook (/Users/Code/woocommerce/node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/runGlobalHook.js:134:15)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async runJest (/Users/Code/woocommerce/node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/runJest.js:369:5)
    at async _run10000 (/Users/Code/woocommerce/node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/cli/index.js:320:7)
    at async runCLI (/Users/Code/woocommerce/node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/cli/index.js:173:3)
    at async Object.run (/Users/Code/woocommerce/node_modules/.pnpm/jest-cli@27.5.1/node_modules/jest-cli/build/cli/index.js:155:37)
```
That said, I don't think that those failing tests need to hold this up because the `e2e-core-tests` could be handled in a follow up PR since no one should be using the trunk versions.

**Comment by tomalec** [comment]
>  tried out the `api-core-tests` and `e2e-core-tests` on the `e2e-environment` package.

Could you tell me more precisely what you did, so I could try to recreate it on my local?

> That said, I don't think that those failing tests need to hold this up because the e2e-core-tests could be handled in a follow up PR since no one should be using the trunk versions.

Does this mean approval? As This PR still needs a formal one, to be merged.
Also, I'm not sure if I could merge with one check failing - https://github.com/woocommerce/woocommerce/runs/8041641423?check_suite_focus=true This seems to be a PHP unit test which is really puzzling to me, as I didn't touch any PHP here.

**Comment by MrJnrman** [comment]
> Could you tell me more precisely what you did, so I could try to recreate it on my local?

From the `plugins/woocommerce` directory, I spun up an environment using `e2e-environment` with `pnpm run docker:up`. Then I started the API tests by running `BASE_URL=http://localhost:8086 USER_KEY=admin USER_SECRET=password DEFAULT_TIMEOUT_OVERRIDE=20000 pnpm exec wc-api-tests test api`

For the e2e tests, I reset the environment by doing `pnpm run docker:down && pnpm run docker:up` and started the tests using `pnpm exec wc-e2e test:e2e` and then I got the error you saw above.


If you can resolve the merge conflict. I'm happy to restest and do a follow up PR on the `e2e-core-tests` to get them healthy. In regards to the failing PHP unit tests, that one is puzzling to me as well. @zhongruige correct me if I'm wrong, but  I have seen that specific set of unit tests fail before and it hasn't blocked merges.

**Comment by zhongruige** [comment]
> but I have seen that specific set of unit tests fail before and it hasn't blocked merges.

Agreed @MrJnrman, and I've also seen some recent failures as well where one test was cancelled after a period of time, and since this doesn't touch the unit tests at all, it should be safe to merge in once the conflicts have been addressed. Plus it may re-run the pipeline with a successful run for that unit test anyway, so we can check the results of it before we merge again.

**Comment by tomalec** [comment]
Thank you @MrJnrman , I merged `trunk` and resolved conflicts.

Running tests on my local produces completely different errors: 
```
  ● Orders API tests › List all orders › pagination

    expect(received).toEqual(expected) // deep equality

    Expected: "10"
    Received: "16"

```

**Comment by tomalec** [comment]
@MrJnrman do you have any hint on how to debug PHP for that test run (debugging JS comes easy)? It does not use my local PHP=>Xdebug, but the docker's one, right? How do I listen to debug from the container?

**Comment by MrJnrman** [comment]
> Running tests on my local produces completely different errors:

Thats pretty odd. I pushed some changes to this branch. Could you pull them and rerun the test after doing a `TURBO_FORCE=true pnpm run build`. This is what I get locally.
Screenshot 2022-09-01 at 9 59 34 AM

> do you have any hint on how to debug PHP for that test run
I don't. These tests are ran against the `@woocommerce/e2e-environment` package which I don't think has support for that. @zhongruige do you know anything about that?

Lastly...I can confirm that tests are being executed locally, but some are currently failing. This isn't a reason to hold this PR since tooling works and the tests themselves need to be updated.

**Comment by zhongruige** [comment]
>  These tests are ran against the @woocommerce/e2e-environment package which I don't think has support for that. @zhongruige do you know anything about that?

I don't believe so. However, we do see issues if the API tests are ran after the end-to-end tests primarily due to using the shared environment (the e2e tests can change locality, currency settings, etc which impacts the test runs--something we're looking at fixing to make them idempotent). So, running them independant of the e2e tests, or bringing down the environment and clearing the Docker containers (such as using `pnpm env:destroy --filter=woocommerce`) can help with some of this flakiness when running locally.

**Comment by MrJnrman** [comment]
Update on the local test run...

I had a total of 8 failed test suites:
Screenshot 2022-09-01 at 3 55 32 PM

Which are:
- `e2e/specs/activate-and-setup/complete-onboarding-wizard.test.js`
- `e2e/specs/admin-homescreen/activity-panel.test.js`
- `e2e/specs/admin-homescreen/task-list.test.js`
- `e2e/specs/admin-tasks/purchase.test.js`
- `e2e/specs/front-end/my-account-pay-order.test.js`
- `e2e/specs/wp-admin/create-shipping-zones.test.js`
- `e2e/specs/wp-admin/create-variable-product.test.js`
- `e2e/specs/wp-admin/page-loads.test.js`
- `e2e/specs/wp-admin/translations.test.js`

**Comment by MrJnrman** [comment]
Looping @woocommerce/solaris on my previous comment since they would own the task of updating those tests.

**Review by MrJnrman — DISMISSED** [review]
Thanks for pushing this forward @tomalec. As said in my comments, API test are passing and E2E tests fail because they need to be updated. 

Just adding a reminder that the [syncpack](https://github.com/woocommerce/woocommerce/pull/34468) should follow closely behind this one.

**Comment by MrJnrman** [comment]
About the failing check. I bypassed creating changelog files because I was prompted to do so after a merge. Those files weren't actually touched in this PR.

**Comment by tomalec** [comment]
> Just adding a reminder that the https://github.com/woocommerce/woocommerce/pull/34468 should follow closely behind this one.

If you are fine with the changes in #34468 we can merge it here, or you can approve it there, so it would be merged just after.

**Comment by tomalec** [comment]
> About the failing check. I bypassed creating changelog files because I was prompted to do so after a merge. Those files weren't actually touched in this PR.

I'm not sure if that means you're fine with this PR as is to be merged?

FYI, I cannot merge it as I'm not authorized.
image

**Review by MrJnrman — APPROVED** [review]
All checks are now passing. @tomalec The changes in https://github.com/woocommerce/woocommerce/pull/34468 look good to me, but I'll wait for the assigned reviewer.

**Pull request merged**
